File: BOJ/practiceBOJ_python/BOJ1197.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

def find(x):
    if x == parent[x]:
        return x
    parent[x] = find(parent[x]) # 부모 테이블 갱신
    return parent[x]

def union(x, y): 
    x = find(x) 
    y = find(y)

    if x == y: # 동일한 집합일 경우
        return

    if x < y:
        parent[y] = x 
    else: 
        parent[x] = y 

V, E = map(int, input().split())
edge = []

for _ in range(E):
    s, e, cost = map(int, input().split())
    edge.append((s, e, cost))

edge.sort(key=lambda x : x[2])

parent = [i for i in range(V+1)]

result = 0
for s, e, c in edge:
    if find(s) != find(e):
        union(s, e)
        result += c

print(result)

File: BOJ/practiceBOJ_java/BOJ1197.java
package study_boj;

import java.io.*;
import java.util.*;

public class BOJ1197 {

	static class Edge implements Comparable<Edge> {
		int s, e, cost;

		Edge(int s, int e, int cost) {
			this.s = s;
			this.e = e;
			this.cost = cost;
		}

		@Override
		public int compareTo(Edge arg0) {
			// Comparable을 통해 정렬 우선순위 (cost 기준)
			return arg0.cost >= this.cost ? -1 : 1;
		}
	}

	static int[] parent;
	static int V, E, result;
	static PriorityQueue<Edge> pque = new PriorityQueue<Edge>();

	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		StringTokenizer st = new StringTokenizer(br.readLine());

		V = Integer.parseInt(st.nextToken());
		E = Integer.parseInt(st.nextToken());

		parent = new int[V + 1];
		for (int i = 1; i <= V; i++) {
			parent[i] = i;
		}

		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());

			int start = Integer.parseInt(st.nextToken());
			int end = Integer.parseInt(st.nextToken());
			int cost = Integer.parseInt(st.nextToken());

			pque.add(new Edge(start, end, cost));
		}

		for (int i = 0; i < E; i++) {
			Edge tmp = pque.poll();

			int a = find(tmp.s);
			int b = find(tmp.e);

			if (a != b) {
				union(a, b);
				result += tmp.cost;
			}
		}
		bw.write(result + "\n");
		bw.flush();
		bw.close();
		br.close();
	}

	// x의 부모 찾기
	public static int find(int x) {
		if (x == parent[x])
			return x;

		return parent[x] = find(parent[x]);
	}

	// y 부모를 x 부모로 치환하기 (x > y 일 경우 반대)
	public static void union(int x, int y) {
		x = find(x);
		y = find(y);

		if (x != y) {
			if (x < y) {
				parent[y] = x;
			} else {
				parent[x] = y;
			}
		}
	}
}
